src/cmdstan-guide/parallelization.Rmd

```diff
@@ -1,37 +1,173 @@
+
+An example of `make/local` on Linux:
+
+```
+STAN_MPI=true
+CXX=mpicxx
+TBB_CXX_TYPE=gcc
+```
+The model is then compiled as normal:
+```
+make path/to/model
+```
+### Running
+
+The Stan model compiled with `STAN_MPI` is then run using an MPI launcher. The MPI standard
+suggests using `mpiexec`, but a vendor wrapper for the launcher like `mpirun` can also be used.
+The launcher is supplied the path to the built executable and the number of processes to start:
+`-n X` for `mpiexec` or `-np X` for `mpirun` where `X` is replaced by the integer representing
+the number of processes.
+
+Example for running a model with six processes:
+```
+mpiexec -n 6 path/to/model sample data file=data.json ...
 ```
 
+## OpenCL
 
+### Dependencies
+
+OpenCL is supported on most modern CPUs and GPUs. In order to run OpenCL-enabled Stan models,
+an OpenCL runtime for the target device must be installed. This subsection lists installation
+instructions for OpenCL runtimes of the commonly-found devices.
+
+In order to check if any OpenCL-enabled device and its runtime is already present use the
+`clinfo` tool. On Linux, `clinfo` can typically be installed with the default package manager
+(for example `sudo apt-get install clinfo` on Ubuntu). For Windows, pre-built `clinfo` binary
+can be found [here](https://github.com/Oblomov/clinfo#windows-support).
+
+Also use `clinfo` to verify successful installation of OpenCL runtimes.
+
+#### NVIDIA GPU
+
+Linux:
+
+Install the NVIDIA GPU driver and the  NVIDIA CUDA Toolkit.
+On Ubuntu the commands to install both is:
+```
+sudo apt update
+sudo apt install nvidia-driver-460 nvidia-cuda-toolkit
+```
+
+Replace the driver version (`460` in the above case) with the lastest number at the time of installation.
+
+Windows:
+
+Install the [NVIDIA GPU Driver](https://www.nvidia.com/Download/index.aspx) and [CUDA Toolkit](https://developer.nvidia.com/cuda-toolkit).
+
+#### AMD GPU
+
+Linux:
+
+Install `Radeon Software for Linux` available [here](https://www.amd.com/en/support/kb/release-notes/rn-amdgpu-unified-linux-20-40).
+
+Windows: 
+
+We recommend installing the open source [OCL-SDK](https://github.com/GPUOpen-LibrariesAndSDKs/OCL-SDK/releases).
+#### AMD CPU
+
+Install the open source [PoCL](http://portablecl.org/download.html).
+
+#### Intel CPU/GPU
+
+Follow Intel's install instructions given [here](https://software.intel.com/content/www/us/en/develop/articles/opencl-drivers.html) (requires registration).
+
+### Compiling
+In order to enable the OpenCL backend the model
+must be compiled with the `STAN_OPENCL` makefile flag. The flag can be 
+supplied in the `make` call as shown below.
+
+```
+make STAN_OPENCL=true path/to/model
+```
+or on Windows
+```
+mingw32-make STAN_OPENCL=true path/to/model.exe
+```
+The flag can also be written to the `make/local` file.
+
+### Running
+
+The Stan model compiled with `STAN_OPENCL` must also be supplied the OpenCL platform and device IDs.
```

Comment:
My computer does defaults and seems to get the right thing.

Reply:
Added a line on defaults working ok.